Add SerenityOS as target

emilytrau opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

A graphical Unix-like operating system for desktop computers! [SerenityOS](https://serenityos.org) is a love letter to '90s user interfaces with a custom Unix-like core. It flatters with sincerity by stealing beautiful ideas from various other systems.

![image](https://user-images.githubusercontent.com/13267947/184506957-510817c7-472f-4219-ae87-68ae4704ad27.png)

I've spent about a week working on porting nixpkgs to SerenityOS. I wanted to play around with both lower-level nixpkgs infra and SerenityOS for the first time and this was a super fun project to do both! There's still plenty of work to be done but I wanted to get it out for play/testing/feedback from both the nix and serenity side. 

Bootstrapping stdenv got very complicated but I'm proud that this builds entirely from source with no bootstrapping binaries required.

This is able to build against the SerenityOS main branch without any patches. I am planning a PR to make the nix profile work out-of-the-box, and start an overlay for Serenity specific package fixes.

:^)

###### How to test

```sh
# Cross-compile a package
nix-build . -A pkgsCross.x86_64-serenity.hello

# Install the closure into your SerenityOS system
nix copy --to $SERENITY_DIR/Build/x86_64/Root/ ./result --no-check-sigs

# Optional: Install a profile for easier access (WIP)
mkdir -p $SERENITY_DIR/Build/x86_64/Root/nix/var/nix/profiles/per-user
nix profile install --profile $SERENITY_DIR/Build/x86_64/Root/nix/var/nix/profiles/per-user/default ./result
nix copy --to $SERENITY_DIR/Build/x86_64/Root/ $SERENITY_DIR/Build/x86_64/Root/nix/var/nix/profiles/per-user/default --no-check-sigs

# Start SerenityOS!
cd $SERENITY_DIR
Meta/serenity.sh run
```

Serenity patch to load the nix profile automatically https://gist.github.com/emilytrau/2c4f8346ca71405527521fbbe0113989

###### Todo

- [x] There is a pre-libc bootstrapping stage where we "manually" compile a GCC cross-compiler toolchain from source. I think this is a bit hacky and ideally it could leverage the existing nixpkgs tooling
- [x] Package SerenityOS's libraries. The toolchain is able to compile simple programs right now, but packages with more complex features (eg. SDL2) will need the libraries to integrate with the OS

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@NixOS/exotic-platform-maintainers 

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ADKaster reviewed (commented):
This is pretty cool! There's already a large collection of patches in the Ports/ directory of serenity, with xexxa's cool https://ports.serenityos.net/ that organizes the [AvailablePorts](https://github.com/SerenityOS/serenity/blob/master/Ports/AvailablePorts.md) by category. 

Your question about SDL2 in the description makes me wonder how Nix would want to pick up or reproduce our existing patches.

emilytrau commented on pkgs/os-specific/serenity/bootstrap-build/default.nix:
```diff
@@ -0,0 +1,109 @@
+    install -D ${tzdb} Build/${serenity_arch}/TZDB/tzdb.tar.gz
+    echo -n ${tzdb_version} > Build/${serenity_arch}/TZDB/version.txt
+    install -D ${tzdb} Build/lagom/TZDB/tzdb.tar.gz
+    echo -n ${tzdb_version} > Build/lagom/TZDB/version.txt
```
Cheers! I've put in a PR to fix this https://github.com/SerenityOS/serenity/pull/14870

emilytrau commented on pkgs/os-specific/serenity/bootstrap-build/default.nix:
```diff
@@ -0,0 +1,109 @@
+    ln -s ${bootstrap-toolchain} Toolchain/Local/${serenity_arch}
+    patchShebangs Meta/
+
+    cmake -S Meta/CMake/Superbuild -B Build/superbuild -GNinja \
```
Figured out why cmakeConfigurePhase was failing and left a comment

emilytrau commented on pkgs/os-specific/serenity/bootstrap-build/default.nix:
```diff
@@ -0,0 +1,109 @@
+  serenity_arch = stdenvNoCC.targetPlatform.parsed.cpu.name;
+
+  tzdb_version = "2022a";
+  tzdb = buildPackages.fetchurl {
```
Cheers

emilytrau commented on pkgs/development/compilers/gcc/12/default.nix:
```diff
@@ -74,6 +75,10 @@ let majorVersion = "12";
       })
       ++ optional langD ../libphobos.patch
+      ++ optionals (targetPlatform.libc == "serenitylibc") (
+        (builtins.filter (lib.hasSuffix ".patch") (lib.filesystem.listFilesRecursive "${serenity.bootstrap-source}/Toolchain/Patches/gcc"))
```
Done

emilytrau commented on lib/tests/systems.nix:
```diff
@@ -29,11 +29,12 @@ with lib.systems.doubles; lib.runTests {
   testfreebsd = mseteq freebsd [ "i686-freebsd" "x86_64-freebsd" ];
   testgenode = mseteq genode [ "aarch64-genode" "i686-genode" "x86_64-genode" ];
   testredox = mseteq redox [ "x86_64-redox" ];
+  testserenity = mseteq serenity [ "i686-serenity" ];
```
This was an easy change that I had just not gotten around to yet. Done :)

emilytrau commented:
Packaged serenity libraries, lagom, and added support for x86_64-serenity. Will squash and organiser properly before merging :)

emilytrau commented:
Not sure why ofborg eval is failing? 😕

emilytrau commented:
@ofborg eval

sternenseemann commented on pkgs/development/compilers/gcc/12/default.nix:
```diff
@@ -75,10 +74,10 @@ let majorVersion = "12";
-      )
+      ++ optionals (targetPlatform.libc == "serenitylibc") [
+        ./serenity-support.patch
+        ./serenity-libtool-enable-shared-library-support.patch
```
cc @aakropotkin for the libtool-related patch

aakropotkin replied:
I'm on mobile and can't view the whole file just the diff.

Just be sure this is in a conditional block tied to the libtool version so it doesn't need to be patched again when this is applied upstream in a new release 🤘

ajs124 replied:
> Just be sure this is in a conditional block tied to the libtool version so it doesn't need to be patched again when this is applied upstream in a new release metal

I think we don't do that in nixpkgs. Probably mainly so we don't carry around patches that aren't needed anymore.
We often apply patches unconditionally, even though they'd only be needed on a specific platform, for the same reason.

aakropotkin replied:
I just read the whole thing and it looks good. There's a chance Serenity users will have issues when libtool 2.8 comes out but that's less destructive than everything being broken.

You could save add a version check to the condition for extra coverage.

alyssais commented on pkgs/development/compilers/gcc/12/default.nix:
```diff
@@ -70,6 +70,10 @@ let majorVersion = "12";
       ++ optional langD ../libphobos.patch
+      ++ optionals (targetPlatform.libc == "serenitylibc") [
+        ./serenity-support.patch
+        ./serenity-libtool-enable-shared-library-support.patch
```
Serenity's Ports directory seems to have copies of this same patch for just about every package using libtool, which worries me. Are we going to need to do that? Why not fix libtool/autoconf/whatever?

emilytrau replied:
In my limited testing the libtool patch doesn't seem to be needed anymore? 🤔 Not sure what might have changed in serenity/autoconf/libtool/etc in the last few months, but definitely welcome if it really has resolved the issue for good!

alyssais reviewed (commented):
I've only reviewed the first 1½ commits so far, because it didn't build for me when I tried, either on the second commit, or the head of the PR:

```
checking for references to /build/ in /nix/store/rh54vhdmfpablzzfjmgnhypbfcs8jq50-serenity-bootstrap-toolchain-unstable-2022-08-07...
RPATH of binary /nix/store/rh54vhdmfpablzzfjmgnhypbfcs8jq50-serenity-bootstrap-toolchain-unstable-2022-08-07/x86_64-pc-linux-gnu/i686-pc-serenity/lib/libopcodes-2.38.so contains a forbidden reference to /build/
find: 'standard output': Broken pipe
find: write error
```

emilytrau committed:
lib: add serenity doubles

emilytrau committed:
stdenv: support serenity

emilytrau committed:
bash: support serenity

emilytrau commented:
Hello friends! I revisited SerenityOS support again and rewrote it from scratch, with changes from all the review comments (thank you!) and upstream support improvements. In my opinion the new patch set is smaller, simpler, and better integrated with nixpkgs compiler tooling. Please let me know your thoughts! @NixOS/exotic-platform-maintainers

Ericson2314 commented:
Is this supposed to be still draft? Would be very cool if we can build the entire system --- good to have pure bootstraps beyond Linux.

emilytrau committed:
gcc: pin version for serenity

alyssais commented on pkgs/top-level/all-packages.nix:
```diff
@@ -40157,6 +40159,10 @@ with pkgs;
     stdenv = crossLibcStdenv;
   };
 
+  serenitylibc = callPackage ../os-specific/serenity/libc.nix {
```
Shouldn't this be serenitylibcCross, and we also have a serenitylibc package without an overridden stdenv, like every other libc?

alyssais commented on pkgs/os-specific/serenity/libc.nix:
```diff
@@ -0,0 +1,105 @@
+          mkdir -p "$(dirname "$out/include/$target")"
+          cp "$header" "$out/include/$target"
+        done
+      '';
```
I'd rather this lived in its own file, and was a proper package with a meta attribute and stuff.

alyssais commented on pkgs/os-specific/serenity/libc.nix:
```diff
@@ -0,0 +1,105 @@
+    [
+      buildPackages.stdenv.cc
+      pkgsBuildBuild.cmake
+      pkgsBuildBuild.ninja
```
Shouldn't this be pkgsBuildHost?  (And why not using splicing?)

emilytrau replied:
`pkgsBuildHost` causes infinite recursion. idk

alyssais replied:
This happens because `stdenv.cc` on Serenity depends on (evaluating) `openssl` for Serenity, and `openssl.separateDebugInfo` depends on `stdenv.cc`.  But `separateDebugInfo` is forced to Linux-only in make-derivation.nix, so breaking the cycle would be easy here — just swap that conditional so that `stdenv.hostPlatform.isLinux` is first, and then it won't need to try evaluating the value at all for non-Linux.

alyssais commented on pkgs/os-specific/serenity/common.nix:
```diff
@@ -0,0 +1,29 @@
+
+    installPhase = ''
+      cp -r . $out
+    '';
```
Why not just attach the meta to the `fetchFromGitHub` directly?  This extra derivation doesn't seem to do anything.

emilytrau replied:
This derivation is entirely for the purposes of running `patchShebangs` on the shell scripts in the repo. I suppose this could be done with `fetchFromGitHub` and adding the hook manually?

alyssais replied:
Ah, apologies, I missed the commment.

alyssais commented on pkgs/development/tools/misc/binutils/default.nix:
```diff
@@ -94,6 +94,9 @@ stdenv.mkDerivation (finalAttrs: {
   # https://github.com/archlinux/svntogit-community/blob/c8d53dd1734df7ab15931f7fad0c9acb8386904c/trunk/avr-size.patch
   ++ lib.optional targetPlatform.isAvr ./avr-size.patch
+  # Adds support for SerenityOS target
+  # https://github.com/SerenityOS/serenity/blob/cff303bd6983cd05848b6f43fb93ec9f346ae9ab/Toolchain/Patches/binutils/0001-Add-support-for-SerenityOS.patch
```
Is the patch modified from Serenity's version at all, or could we use fetchurl?

alyssais commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -224,30 +225,42 @@ in {
+      ++ lib.optionals stdenv.hostPlatform.isSerenity ([
+        ./support-serenity.patch
+        ./3.0/serenity-os-data.patch
+      ]);
```
parens here are unnecessary

alyssais commented on pkgs/development/compilers/gcc/common/configure-flags.nix:
```diff
@@ -229,6 +229,14 @@ let
+      "--with-linker-hash-style=gnu"
+      "--disable-nls"
+      # provided by libc
+      "--disable-libssp"
```
Maybe this could be merged into the NetBSD case above? Unsure whether that would be better.

alyssais reviewed (changes requested):
This is looking much better.  Thanks for your continued work on it!

The main concern I have now is the patches.  You're applying them all conditionally, which is good in the sense that it won't annoy people updating packages, but bad for actually using Nixpkgs with Serenity, because it means packages are likely to bitrot.  It's also not usual Nixpkgs practice to conditionally apply patches like this.

The best way to avoid this tradeoff would be for the patches to go upstream.  What's the status of upstreaming them?

emilytrau commented:
@alyssais This issue tracks the status of upstreaming SerenityOS support patches https://github.com/SerenityOS/serenity/issues/14183

emilytrau committed:
serenitylibc: rename to serenitylibcCross

emilytrau committed:
binutils: fetch serenity patch

emilytrau committed:
gcc: format serenity flags

emilytrau committed:
serenity: package libraries

emilytrau committed:
serenity: update to gcc 13

emilytrau committed:
serenity.serenity-headers: refactor into own file

emilytrau commented on pkgs/os-specific/serenity/default.nix:
```diff
@@ -0,0 +1,108 @@
+      inherit pname version;
+
+      # FIXME: rpath is not rewritten to use dependency derivations
+      buildInputs = builtins.attrValues deps;
```
This is supposed to drop direct references to `serenity-rootfs` and replace them with dependencies defined in  `libraries.nix`. I've tried combinations of `patchelf` and `autoPatchelfHook` but without success

emilytrau committed:
serenity: regenerate libraries

emilytrau committed:
serenity: remove references to rootfs from libraries

emilytrau committed:
stdenv: enable default pie on serenity

emilytrau committed:
serenity.rootfs: support cross-compiling from darwin

emilytrau committed:
maintainers: add serenity team

alyssais commented:
> @alyssais This issue tracks the status of upstreaming SerenityOS support patches [SerenityOS/serenity#14183](https://github.com/SerenityOS/serenity/issues/14183)

This doesn't fill me with confidence, because it doesn't look like anything at all has been upstreamed in the last year, and the only stuff ever upstreamed is the GNU config stuff.  How can we be confident we're not going to need to keep all these patches around forever?

emilytrau committed:
lib.systems: add aarch64-serenity double

emilytrau committed:
gcc: unvendor serenity patches

emilytrau committed:
serenity.rootfs: use splicing for dependencies

emilytrau commented:
@alyssais sorry i brainfarted somehow and forgot we could also fetch the gcc and openssl patches :sweat_smile: 

also i was THRILLED to find out that `aarch64-serenity` just!works! after adding the double! :DD

siraben commented:
What’s the status of this PR? IMO having diverse cross targets to systems both hobbyist and commercial is a great way to increase the usefulness of Nix, as long as there aren’t changes that affect darwin and linux.

alyssais commented:
The status of this PR from my point of view is that it's ready to go, except that I don't want to see Nixpkgs fill up with patches (checked into the repo or not) that aren't going to ever go away.  If they had been submitted upstream and rejected, that would be different, because some upstreams would accept Serenity patches, and others wouldn't, and overall we wouldn't end up with too many.  But the problem here is that as far as I can tell, the Serenity folks haven't even attempted to upstream their patches, and that means that supporting Serenity will end up needing to apply lots of patches to lots of packages, none of which will ever go away unless something changes and they start trying to upstream them.  Maintaining a large number of packages with platform-specific conditionally-applied patches in Nixpkgs is no fun (I've done it), and I think that if we accept that approach, the platform support will end up being too difficult/annoying to maintain.

I really want to see Serenity support in Nixpkgs, but what I need to see to be happy with this PR is some movement on the Serenity side towards upstreaming their patches.  I understand that it hasn't been a big priority until now, because all the patches are only used in one place (either the Serenity build system or ports tree), but with Nixpkgs being a second provider of packages for Serenity, that changes, and getting the patches upstream becomes a lot more important.  I don't mind patching where upstreams reject Serenity support, but I don't think it would be sustainable to support a platform where upstreaming isn't even attempted.

@networkException recently asked me about the status of this PR, and told me (IIRC) that using LLVM instead of GCC for the Serenity platform would avoid the need for compiler patches.  So if nobody can be convinced to upstream the GCC patches, maybe that's an alternative, but it still doesn't solve the bigger issue that we should be asking "has this been submitted upstream?" every time somebody wants to apply a patch in Nixpkgs (not just for Serenity).

ADKaster commented:
@alyssais I created an RFC to upstream the patches we have to LLVM here https://discourse.llvm.org/t/rfc-add-support-for-serenityos/71641

Hopefully after we get at least one of the compilers' patches upstreamed, we can work to get the rest of the Ports' patches upstreamed.

emilytrau commented:
Glad to hear this project has helped motivate the upstreaming effort! In the mean time I will work on getting the LLVM toolchain to work as well. To focus on the compiler toolchain I've dropped the patches for OpenSSL.

emilytrau committed:
serenity: unstable-2023-05-18 -> unstable-2023-06-27

Last revision before major LLVM bump https://github.com/SerenityOS/serenity/pull/18522

ehmry commented:
We should consider keeping enough patches for Serenity in-tree to make stdenv work but not patches and conditionals outside of that scope. I think there is less work overall if Serenity testers don't have to build their own toolchains and if we don't have to work around their conditionals in the leaf packages here.

Accumulating patches for toolchains isn't such a terrible thing. It would be a resource for people looking to port toolchains elsewhere and if there are problems at least its not something that we deal with on the master branch.

alyssais commented:
> Accumulating patches for toolchains isn't such a terrible thing.

Agreed, which is why I'm happy to accept them if they're rejected upstream.  But it's definitely not _better_ than the patches being upstreamed, which is why we should only do it if they can't be upstreamed.

alyssais commented:
> Maybe the platform only wants to support building with GCC and not LLVM.

This is not the case.

alyssais commented:
> patching their master branch and waiting for the next release can take a while.

There's a major LLVM release [every six months](https://www.llvm.org/docs/HowToReleaseLLVM.html).  That's not a long time in the context of adding a whole new platform.  And it would be fine to backport patches to get support available early — backports are very different to patches we might never get rid of.

> Maybe the platform only wants to support building with GCC and not LLVM.

In that case, it would be fine not to support LLVM for that platform.

ehmry commented:
A release every six months means waiting three months on average, plus the time to get patches merged. This PR has been open for ten months already and I would not be surprised if it was open in ten months from now.

alyssais commented:
> Can we please have an example where we've been burned in the past by this?

Example one: I maintain QEMU.  For a long time updating QEMU was a total nightmare, because of Darwin patches that were applied in Nixpkgs, and those were just the bare minimum to deal with QEMU depending on a newer SDK than we are able to provide.  Updates would be delayed weeks because of the need to update patches, and because they were for Darwin, I couldn't even test them.  Later there was a proposal to add **eleven** more patches to implement 9p support on Darwin.  I put my foot down, because at that point there's no way I'd be able to handle updates without them getting blocked on Darwin.  As a result, the patches were submitted and accepted upstream, and now every QEMU user on Darwin, not just Nixpkgs users, benefits from that support, and I don't have to try to rebase a dozen patches for a platform I don't use every time I update QEMU.

Example two: we patch systemd to work on musl, because upstream is hostlie to musl.  As a result of a compromise with the systemd maintainers, these patches are only applied during musl builds, so they don't have to wait for updated musl patches to update systemd.  (This is the same setup originally proposed for the Serenity toolchain.)  This _works_, but it means that as musl enthusiasts, we have to be very on the ball so that new systemd releases don't completely break the ability to run NixOS on musl.  (Basically, the patches have to be updated before systemd gets out of staging.)  A lot of work would have been saved if upstream had been willing to accept those patches once, so they didn't have to be redone every time.  Sadly, in this case that didn't happen, so we're stuck with extra boring, repetitive work forever keeping those patches up to date.  It was absolutely worth the attempt to upstream the patches to try to save all that future work, especially because, if upstream is entirely unwilling to support a platform, it's probably not going to take very long to find that out.

alyssais commented:
> It is immensely easier to say "no more patches"

Nobody is saying more patches.  I'm saying "no patches where there hasn't even been the bare minimum  of an attempt to upstream them, where they would make sense to upstream, and upstream is not known to be unreasonable (e.g. hostile to new platforms, bad CLA requirements, etc.)".

siraben commented:
@emilytrau are you still planning on merging this? This would be a cool addition to the cross infra!
